### PR TITLE
Paste product combination tokens and highlight invalid tokens

### PR DESCRIPTION
**Pull request by MarkALeonard**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Allow product combination tokens to be pasted. Pasted comma separated text will be validated and tokens added. Invalid tokens will be coloured yellow. Duplicate tokens will not be added.
| Type?         |  improvement 
| Category?     |  BO
| BC breaks?    |  no
| Deprecations? |  no
| Fixed ticket? | Fixes #{issue number here}.
| How to test?  | **On Current Product Page:**  Copy and paste 'Color : Red, Color : Blue, Color : Red, Color : Green, Color : Invalid' into the product combination tokens field. All tokens will be created including duplicates but no attributes will be selected and pressing the generate button produces a single product combination. **On Product Page with this Improvement:** Copy and paste 'Color : Red, Color : Blue, Color : Red, Color : Green, Color : Invalid' into the product combination tokens field. Tokens will be created, duplicate tokens will be removed, attributes will be selected and invalid tokens highlighted in yellow. Pressing the generate button will create the correct number of product combinations.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/18095)

### CODE REVIEW

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 1]
```diff
@@ -929,19 +929,51 @@ var form = (function() {
...
+            engine.search(e.attrs.label, function(result) {
+              if (result.length == 1) {
+                e.attrs.label = result[0].label;
+                e.attrs.value = result[0].value;
+                e.attrs.data = [];
+                e.attrs.data['id_group'] = result[0].data.id_group;
+              }
+            });
+          } else {
+            var attr = $('.js-attribute-checkbox[data-value="' + e.attrs.value + '"]');
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
            var attr = $(`.js-attribute-checkbox[data-value="${e.attrs.value}"]`);
```

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 2]
```diff
@@ -929,19 +929,51 @@ var form = (function() {
...
+
+            if (attr != undefined) {
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
            if (attr) {
```

Wouldn't be just enough ? Else prefer using strict operators anyway

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 3]
```diff
@@ -929,19 +929,51 @@ var form = (function() {
...
+              e.attrs.label = attr.data('label');
+              e.attrs.value = attr.data('value');
+              e.attrs.data = [];
+              e.attrs.data['id_group'] = attr.data('group-id');
+            }
+          }
+
+          if(e.attrs.data){
+            if(filter([e.attrs]).length == 0){
+              $('#form_step3_attributes-tokenfield').val(function(i, value){
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js]: I've a doubt, this file is processed with webpack right? If yes then prefer using arrow functions
```suggestion
              $('#form_step3_attributes-tokenfield').val((i, value) => {
```

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

**Comment by NeOMakinG** [comment]
Thanks for your contribution @MarkALeonard, I left some comments mainly aiming good practices :)

**Comment by MarkALeonard** [comment]
Thanks @NeOMakinG for your feedback, I have made the changes accordingly.

**Comment by NeOMakinG** [comment]
> Thanks @NeOMakinG for your feedback, I have made the changes accordingly.

Did you test it after changes ?  Hope the `if (attr)` is fine in most case, thanks for you work.

**Review by NeOMakinG — DISMISSED** [review]

**Comment by MarkALeonard** [comment]
Yes, tested and works as expected. Thank you.

**Comment by MarkALeonard** [comment]
On implementing on a live site I noticed the Bloodhound engine returns multiple results for 'Size : S' but not 'Size : M'. Have changed `result.length == 1` to `result.length >= 1` and first result in array is used. Also fixed issue from original code that reported an error when removing invalid tokens. As the new code has the 'invalid' class I have added a check for that. Apologies for not noticing this initially.

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 4]
```diff
@@ -929,25 +929,59 @@ var form = (function() {
...
       };
 
-      /** On event "tokenfield:createtoken" : stop event if its not a typehead result */
+      /** On event "tokenfield:createtoken" : check values are valid if its not a typehead result */
       $('#form_step3_attributes').on('tokenfield:createtoken', function(e) {
         if (!e.attrs.data && e.handleObj.origType !== 'tokenfield:createtoken') {
           return false;
         }
+
+        if (!e.attrs.data){
```
Comment (Progi1984) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
        if (!e.attrs.data)
          if (e.handleObj.origType !== 'tokenfield:createtoken') {
            return false;
          }
```

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 5]
```diff
@@ -929,25 +929,58 @@ var form = (function() {
...
+            if (attr) {
+              e.attrs.label = attr.data('label');
+              e.attrs.value = attr.data('value');
+              e.attrs.data = [];
+              e.attrs.data['id_group'] = attr.data('group-id');
+            }
+          }
+
+          if(e.attrs.data){
+            if(filter([e.attrs]).length == 0){
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
            if(filter([e.attrs]).length === 0){
```

In that case that's unnecessary but as we want to keep consistance in our codebase, can you prefer using strict operators ?

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 6]
```diff
@@ -929,25 +929,58 @@ var form = (function() {
...
+      /** On event "tokenfield:createtoken" : check values are valid if its not a typehead result */
       $('#form_step3_attributes').on('tokenfield:createtoken', function(e) {
-        if (!e.attrs.data && e.handleObj.origType !== 'tokenfield:createtoken') {
-          return false;
+        if (!e.attrs.data)
+          if (e.handleObj.origType !== 'tokenfield:createtoken') {
+            return false;
+          }
+          var orgLabel = e.attrs.label;
+          if (e.attrs.label == e.attrs.value) {
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
          if (e.attrs.label === e.attrs.value) {
```

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 7]
```diff
@@ -929,25 +929,58 @@ var form = (function() {
...
+        } else {
+          $(e.relatedTarget).addClass('invalid');
         }
       });
 
       /** On event "tokenfield:removedtoken" : remove stored attributes input when remove token */
       $('#form_step3_attributes').on('tokenfield:removedtoken', function(e) {
-        $('#attribute-generator-' + e.attrs.value).remove();
+        if(!$(e.relatedTarget).hasClass('invalid')){
+          $('#attribute-generator-' + e.attrs.value).remove();
```
Comment (NeOMakinG) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
          $(`#attribute-generator-${e.attrs.value}`).remove();
```

**Review by NeOMakinG — CHANGES_REQUESTED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Review by NeOMakinG — DISMISSED** [review]
Thanks!

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 8]
```diff
@@ -929,25 +929,59 @@ var form = (function() {
...
 
-      /** On event "tokenfield:createtoken" : stop event if its not a typehead result */
+      /** On event "tokenfield:createtoken" : check values are valid if its not a typehead result */
       $('#form_step3_attributes').on('tokenfield:createtoken', function(e) {
-        if (!e.attrs.data && e.handleObj.origType !== 'tokenfield:createtoken') {
-          return false;
+        if (!e.attrs.data)
+          if (e.handleObj.origType !== 'tokenfield:createtoken') {
+            return false;
+          }
```
Comment (PierreRambaud) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
        if (!e.attrs.data && e.handleObj.origType !== 'tokenfield:createtoken') {
          return false;
        }
```

Comment (PierreRambaud) [reply admin-dev/themes/default/js/bundle/product/form.js]: Could you explain why you split the if statement?

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 9]
```diff
@@ -929,25 +929,59 @@ var form = (function() {
...
+          
+        var orgLabel = e.attrs.label;
```
Comment (PierreRambaud) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
        const orgLabel = e.attrs.label;
```

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 10]
```diff
@@ -929,25 +929,59 @@ var form = (function() {
...
+          engine.search(e.attrs.label, function(result) {
+            if (result.length >= 1) {
+              e.attrs.label = result[0].label;
+              e.attrs.value = result[0].value;
+              e.attrs.data = [];
+              e.attrs.data['id_group'] = result[0].data.id_group;
+            }
+          });
+        } else {
+          var attr = $(`.js-attribute-checkbox[data-value="${e.attrs.value}"]`);
```
Comment (PierreRambaud) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
          const attr = $(`.js-attribute-checkbox[data-value="${e.attrs.value}"]`);
```

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 11]
```diff
@@ -929,25 +929,59 @@ var form = (function() {
...
         }
+
+        if(e.attrs.data){
+          if(filter([e.attrs]).length === 0){
+            $('#form_step3_attributes-tokenfield').val((i, value) => {
+              return value.replace(orgLabel,"");
+            });
+            return false;
+          };
+        }
```
Comment (PierreRambaud) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
      if(e.attrs.data && filter([e.attrs]).length === 0){
          $('#form_step3_attributes-tokenfield').val((i, value) => {
            return value.replace(orgLabel,"");
          });
          return false;
        }
```

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 12]
```diff
@@ -929,25 +929,59 @@ var form = (function() {
...
-          $('#attributes-generator').append('<input type="hidden" id="attribute-generator-' + $('.js-attribute-checkbox[data-value="'+e.attrs.value+'"]').data('value') + '" class="attribute-generator" value="' + $('.js-attribute-checkbox[data-value="'+e.attrs.value+'"]').data('value') + '" name="options[' + $('.js-attribute-checkbox[data-value="'+e.attrs.value+'"]').data('group-id') + '][' + $('.js-attribute-checkbox[data-value="'+e.attrs.value+'"]').data('value') + ']" />');
+        } else {
+          $(e.relatedTarget).addClass('invalid');
         }
       });
 
       /** On event "tokenfield:removedtoken" : remove stored attributes input when remove token */
       $('#form_step3_attributes').on('tokenfield:removedtoken', function(e) {
-        $('#attribute-generator-' + e.attrs.value).remove();
+        if(!$(e.relatedTarget).hasClass('invalid')){
```
Comment (PierreRambaud) [inline admin-dev/themes/default/js/bundle/product/form.js]:
```suggestion
        if (!$(e.relatedTarget).hasClass('invalid')) {
```

File: admin-dev/themes/default/js/bundle/product/form.js (hunk 8 above)
Comment (MarkALeonard) [reply admin-dev/themes/default/js/bundle/product/form.js]: The statement was split as the next section of the code only needs to run if the data attributes have not been set. The revision was suggested earlier in this converstion. However I have noticed that with the change including the dropping of braces it actually doesn't work as intended. I will restructure it and push a new version which will make the decision clearer.

**Comment by khouloudbelguith** [comment]
Hi @MarkALeonard,

Can you please resolve conflicts.

Thanks!

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Comment by SD1982** [comment]
LGTM Thanks @MarkALeonard

**Comment by Progi1984** [comment]
Thanks @MarkALeonard

**Pull request merged**